### PR TITLE
White Label: can upload a custom header logo

### LINKED ISSUE #10559: Allow upload of custom header logo and url

### PR DESCRIPTION
**Pull request by jibees**
#### What? Why?

- Closes #10559
⚠️ Behind a feature toggle `white_label`

This PR introduce a new preference:  
Capture d’écran 2023-03-17 à 15 21 37

And then it's used in the same page than the other [PR](https://github.com/openfoodfoundation/openfoodnetwork/pull/10567): `shop`, `cart`, `checkout` and `confirmation page`. It basically remplace the logo by the uploaded image for both desktop and mobile view. 

#### What should we test?

- As an admin, activate `Hide OFN Navigation` under the `White Label` panel for an enterprise
- Upload a new image
- See that on `shop`, `cart`, `checkout` and `confirmation page` for the distributor, the logo is present and the OFN one is not show.
- If the feature is not activated, nothing should change.

#### Release notes

Changelog Category: User facing changes

### CODE REVIEW

**Comment by jibees** [comment]
@filipefurtad0 
There are some `xit` in my PR, since I'm don't know how to handle image attachmement for a distributor (`white_label_logo`) in spec. 

For example, I want to test the case the distributor already has white_label_logo:

```ruby
context "when enterprise has a white label logo" do
  before do
    distributor1.update white_label_logo: white_logo_file
  end

  xit "can remove the white label logo for the current shop" do
    click_button "Remove"
    within ".reveal-modal" do
      click_button "Confirm"
    end
    expect(flash_message).to eq("Logo removed")
    expect(distributor1.reload.white_label_logo).to be_nil
  end
end
```

Unfortunately, this doesn't seem to work like this. Can you please help me a bit? Many thanks! 🙏

File: spec/system/admin/enterprises_spec.rb
Comment (filipefurtad0) [inline spec/system/admin/enterprises_spec.rb:595]: Hey @jibees ,

I was able to attach a file with:
```suggestion
          context "when enterprise has a white label logo" do
                      let(:white_logo_file) { Rails.root.join("public/default_images/ofn-logo.png") }

            before do
              attach_file "enterprise[white_label_logo]", white_logo_file
            end
```
After this change, we can see the file is attached (by running with the headless mode off), but the `Remove` button does not seem to appear... Maybe some code change is still needed?

Comment (jibees) [reply spec/system/admin/enterprises_spec.rb:595]: Sorry, I should have been clearer ;) 
I can attach a file to a the form:
https://github.com/jibees/openfoodnetwork/blob/10559-allow-upload-of-custom-header-logo/spec/system/admin/enterprises_spec.rb#L589

but what I'm not able to do is to set to a distributor a white label logo. 
I've the same issue here: 
https://github.com/jibees/openfoodnetwork/blob/10559-allow-upload-of-custom-header-logo/spec/system/consumer/white_label_spec.rb#L221 

(that's why I've a `xit` few lines after). 

So my question is: how can I set a white label logo (defined has an attachment [here](https://github.com/jibees/openfoodnetwork/blob/10559-allow-upload-of-custom-header-logo/app/models/enterprise.rb#L91)) to an enterprise in specs? 

Thanks! 🙏

Comment (filipefurtad0) [reply spec/system/admin/enterprises_spec.rb:595]: Oh, I see now - sorry, I misunderstood.

I've looked at https://github.com/jibees/openfoodnetwork/blob/10559-allow-upload-of-custom-header-logo/spec/system/consumer/white_label_spec.rb#L221 now.

I think the syntax is fine, and that the logo can be set in this way. However, I've noticed that the dimensions of the `app/webpacker/images/logo-white.png` are not the same as `/default_images/ofn-logo.png`. This seems to make the image not visible:

image

But this is not due to syntax. If we try the same syntax for a file with correct dimensions - for example `app/webpacker/images/logo-australia.png` then it appears:

image

All this should actually not really matter, as we are only asserting on the path - sorry for the detour :sweat_smile: But I was not sure we were able to upload files this way.

Ok, so now, knowing that images uploaded (i.e., updated on the correct `distributor`) successfully, the assertion on the path fails:

`expect(page).to have_selector "img[src*='#{white_logo_file}']"`

It only worked if I replaced the line above with:

`expect(page).to have_selector "img[src*='/logo-white.png']"`

Is this helpful?

Comment (jibees) [reply spec/system/admin/enterprises_spec.rb:595]: Nice, I think it's perfect now ;)
Thanks for the help, I missed that because I cannot see the image when running in non-headless mode, but it's because it's a white logo on a white nav bar! So _invisible_... Noob I was. But anyway, syntax was almost correct ;)

🙏

File: app/assets/javascripts/admin/admin_ofn.js.coffee
Comment (dacook) [inline app/assets/javascripts/admin/admin_ofn.js.coffee:15]: This commit doesn't move or change anything, has it since been rebased? 

Still, the comment seems useful

File: app/models/enterprise.rb
Comment (dacook) [inline app/models/enterprise.rb]: I can see this is the rendered size of the existing logo on desktop, so it seems like the ideal resolution to store.

Many screens can display at higher density (eg macbook and most phones). But it can be a bit of a rabbit-hole trying to optimise all this so let's just stick with this for now ;)

Comment (dacook) [reply app/models/enterprise.rb]: > `# The next Rails version will have named variants`

It could have been good to try that out 🤷

File: app/views/admin/enterprises/form/_white_label.html.haml
Comment (dacook) [inline app/views/admin/enterprises/form/_white_label.html.haml:15]: 👀 First time I've seen `variable?`. I can see it is used to determine the image's presence elsewhere in the codebase, but I can't find any info about this method online!

File: spec/system/admin/enterprises_spec.rb
Comment (dacook) [inline spec/system/admin/enterprises_spec.rb]: `±` Actually, `hide_ofn_navigation` doesn't need to be set before you can add/remove the logo image, does it? 

So this context seems unnecessary.

File: app/webpacker/controllers/checkbox_display_controller.js
Comment (dacook) [inline app/webpacker/controllers/checkbox_display_controller.js:29]: `±` 
`targetValue` is an ID, so I would rename to `targetId`

File: app/assets/javascripts/admin/admin_ofn.js.coffee
Comment (mkllnk) [inline app/assets/javascripts/admin/admin_ofn.js.coffee:17]: Has the fix been merged independently? This commit is just adding comments now.

File: app/models/enterprise.rb
Comment (mkllnk) [inline app/models/enterprise.rb]: Rails 7 needs these options in a different format.

* https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#active-storage-default-variant-processor-changed-to-vips

Well, I think that we are not using vips yet and therefore it's working now but it will break when we change.

**Review by mkllnk — CHANGES_REQUESTED** [review]
Great work. Just one update needed.

And I would keep this PR In Dev until the blocking PR is merged.

**Review by dacook — CHANGES_REQUESTED** [review]
This is following on from https://github.com/openfoodfoundation/openfoodnetwork/pull/10567 which hasn't been merged yet, so ideally we wouldn't review this yet. 

But I had a look at the new commits anyway. Looks good, I just have a request to update a variable name.

File: app/views/admin/enterprises/form/_white_label.html.haml
Comment (mkllnk) [reply app/views/admin/enterprises/form/_white_label.html.haml:15]: https://api.rubyonrails.org/v6.1.7.3/classes/ActiveStorage/Blob/Representable.html#method-i-variable-3F

Comment (mkllnk) [reply app/views/admin/enterprises/form/_white_label.html.haml:15]: Some images can be resized, others can't.

Comment (dacook) [reply app/views/admin/enterprises/form/_white_label.html.haml:15]: So we only display images that can be resized (or otherwise reprocessed)? Seems like an indirect way of checking for a valid image 🤷

File: app/models/enterprise.rb
Comment (jibees) [reply app/models/enterprise.rb]: Changed: 2b0d3a6134

Comment (jibees) [reply app/models/enterprise.rb]: Changed! 2b0d3a6134

File: app/assets/javascripts/admin/admin_ofn.js.coffee
Comment (jibees) [reply app/assets/javascripts/admin/admin_ofn.js.coffee:15]: Yep: https://github.com/openfoodfoundation/openfoodnetwork/pull/10575#discussion_r1151916464

File: app/views/admin/enterprises/form/_white_label.html.haml
Comment (jibees) [reply app/views/admin/enterprises/form/_white_label.html.haml:15]: Changed to `.present?`

File: spec/system/admin/enterprises_spec.rb
Comment (jibees) [reply spec/system/admin/enterprises_spec.rb]: > Actually, hide_ofn_navigation doesn't need to be set before you can add/remove the logo image, does it?

You can't add any logo if `hide_ofn_navigation` is not true.

File: app/webpacker/controllers/checkbox_display_controller.js
Comment (jibees) [reply app/webpacker/controllers/checkbox_display_controller.js]: I've changed to `target-id` (in html), `targetId` (in stimulus controller) and then `targetIdValue` (to get its value). 

--> https://stimulus.hotwired.dev/reference/values

**Comment by jibees** [comment]
I spend much time to rebase each PR on others. I think I should wait for others to be merged before reviewing them. That's why this one is sill in `In Dev` status.

File: app/views/admin/enterprises/form/_white_label.html.haml
Comment (mkllnk) [reply app/views/admin/enterprises/form/_white_label.html.haml:15]: Good, it wasn't the right check here before. The `variable?` is important before trying to display an image as a certain variant because that will just raise an error if it's not variable.

File: spec/system/admin/enterprises_spec.rb
Comment (dacook) [reply spec/system/admin/enterprises_spec.rb]: Sorry, I missed that when I made this comment!

**Comment by jibees** [comment]
Since #10567 has been merge, I rebased since one onto master, and then moved in `Code Review`. 
This one is ready to be reviewed!

File: app/models/enterprise.rb
Comment (mkllnk) [inline app/models/enterprise.rb:37]: This seems to be a leftover from the old branch. You got an attribute on the enterprise now and you don't need this, right?

File: app/assets/javascripts/admin/admin_ofn.js.coffee
Comment (mkllnk) [reply app/assets/javascripts/admin/admin_ofn.js.coffee:17]: Yes, the comment is useful but I would change the commit message to reflect what's happening.

**Review by mkllnk — CHANGES_REQUESTED** [review]
Cool, the variants are looking better now. You don't have to do it in this pull request but with Rails 7 we have a better option for names variants: https://stackoverflow.com/questions/72238951/rails-active-storage-how-to-create-named-variants-that-are-cropped-by-user-su

I think the PR needs a bit more updating after the previous one was changed.

File: app/models/enterprise.rb
Comment (jibees) [reply app/models/enterprise.rb:37]: Good catch, commit removed.

**Comment by jibees** [comment]
@mkllnk 
>You don't have to do it in this pull request but with Rails 7 we have a better option for names variants

I made the choice to postpone it. 


I removed a commit, change the commit message for another one. I hope this one is ready now!

File: app/reflexes/white_label_reflex.rb
Comment (dacook) [inline app/reflexes/white_label_reflex.rb:16]: I didn't pick up on this before, but just noticed that this part is simply re-rendering the white_label form area. Excuse me while I muse...

This seems like it would be a common pattern for reflexes, ie:
1. make a change
2. re-render the form
3. display a response

For re-rendering the form, we could probably add some structure around it. To begin with, I'd made it a separate method (it could be re-used by other reflexes in the future).

But I can also imagine in the future having more 'rails magic' that works by default, just like a rails controller knows how to render a view.

Right now, I don't suggest changing, but I think we should keep looking for ways to add helpful structure to the codebase..

**Review by dacook — APPROVED** [review]
Looks good to me!

File: app/reflexes/white_label_reflex.rb
Comment (mkllnk) [reply app/reflexes/white_label_reflex.rb:16]: Ah, so you mean that we could have a helper like this:

```
    morph_element "white_label_panel", locals: { f: f, enterprise: @enterprise }
```

And it takes the first parameter as `id` and partial name. The rest is passed on to `render`? There may be other parameters we would need to consider but the idea is good. Hm, or should we just use component reflexes then?

**Review by mkllnk — APPROVED** [review]

File: app/reflexes/white_label_reflex.rb
Comment (dacook) [reply app/reflexes/white_label_reflex.rb:16]: >  Hm, or should we just use component reflexes then?

I forgot about that! I need to get into this some more.

**Comment by mkllnk** [comment]
Failing spec. CI says:
```
ailures:

  1) 
    As an administrator
    I want to manage complex order cycles
 editing an order cycle with multiple pages of products select all products
     Failure/Error: page.execute_script("document.getElementById('#{checkbox_id}').scrollIntoView()")
     
     Ferrum::JavaScriptError:
       TypeError: Cannot read properties of null (reading 'scrollIntoView')
           at <anonymous>:1:93
     
     [Screenshot Image]: /home/runner/work/openfoodnetwork/openfoodnetwork/tmp/capybara/screenshots/failures_r_spec_example_groups_as_an_administrator_i_want_to_manage_complex_order_cycles_editing_an_order_cycle_with_multiple_pages_of_products_select_all_products_358.png

     
     # ./spec/system/admin/order_cycles/complex_editing_multiple_product_pages_spec.rb:40:in `block (3 levels) in <top (required)>'
     # ./spec/system/support/cuprite_setup.rb:44:in `block (2 levels) in <top (required)>'

Finished in 3 minutes 18.2 seconds
71 examples, 1 failure

Failed examples:

rspec ./spec/system/admin/order_cycles/complex_editing_multiple_product_pages_spec.rb:37
```

My local machine says:

```
Failures:

  1) 
    As an administrator
    I want to manage complex order cycles
 editing an order cycle with multiple pages of products select all products
     Failure/Error: visit path_visit
     
     Ferrum::JavaScriptError:
       Error: Module build failed (from ./node_modules/babel-loader/lib/index.js):
       TypeError: /home/maikel/code/openfoodnetwork/app/webpacker/controllers/checkbox_display_controller.js: isTSSatisfiesExpression is not a function
           at isTSTypeExpression (/home/maikel/code/openfoodnetwork/node_modules/@babel/core/node_modules/@babel/generator/lib/node/parentheses.js:113:36)
           at ConditionalExpression (/home/maikel/code/openfoodnetwork/node_modules/@babel/core/node_modules/@babel/generator/lib/node/parentheses.js:207:38)
           at AssignmentExpression (/home/maikel/code/openfoodnetwork/node_modules/@babel/core/node_modules/@babel/generator/lib/node/parentheses.js:223:12)
           at find (/home/maikel/code/openfoodnetwork/node_modules/@babel/core/node_modules/@babel/generator/lib/node/index.js:45:15)
           at needsParens (/home/maikel/code/openfoodnetwork/node_modules/@babel/core/node_modules/@babel/generator/lib/node/index.js:75:10)
           at Generator.print (/home/maikel/code/openfoodnetwork/node_modules/@babel/core/node_modules/@babel/generator/lib/printer.js:325:27)
           at Generator.ExpressionStatement (/home/maikel/code/openfoodnetwork/node_modules/@babel/core/node_modules/@babel/generator/lib/generators/expressions.js:218:8)
           at Generator.exactSource (/home/maikel/code/openfoodnetwork/node_modules/@babel/core/node_modules/@babel/generator/lib/printer.js:155:7)
           at Generator.print (/home/maikel/code/openfoodnetwork/node_modules/@babel/core/node_modules/@babel/generator/lib/printer.js:334:10)
           at Generator.printJoin (/home/maikel/code/openfoodnetwork/node_modules/@babel/core/node_modules/@babel/generator/lib/printer.js:395:12)
           at Object../app/webpacker/controllers/checkbox_display_controller.js (http://127.0.0.1:33725/packs-test/js/admin-49f8f69ee0faf6da4292.js:539:7)
           at __webpack_require__ (http://127.0.0.1:33725/packs-test/js/admin-49f8f69ee0faf6da4292.js:20:30)
           at webpackContext (http://127.0.0.1:33725/packs-test/js/admin-49f8f69ee0faf6da4292.js:165:9)
           at definitionForModuleWithContextAndKey (http://127.0.0.1:33725/packs-test/js/admin-49f8f69ee0faf6da4292.js:7102:45)
           at http://127.0.0.1:33725/packs-test/js/admin-49f8f69ee0faf6da4292.js:7092:12
           at Array.map (<anonymous>)
           at definitionsFromContext (http://127.0.0.1:33725/packs-test/js/admin-49f8f69ee0faf6da4292.js:7091:25)
           at Module../app/webpacker/packs/admin.js (http://127.0.0.1:33725/packs-test/js/admin-49f8f69ee0faf6da4292.js:4452:105)
           at __webpack_require__ (http://127.0.0.1:33725/packs-test/js/admin-49f8f69ee0faf6da4292.js:20:30)
           at http://127.0.0.1:33725/packs-test/js/admin-49f8f69ee0faf6da4292.js:84:18
     
     [Screenshot Image]: /home/maikel/code/openfoodnetwork/tmp/capybara/screenshots/failures_r_spec_example_groups_as_an_administrator_i_want_to_manage_complex_order_cycles_editing_an_order_cycle_with_multiple_pages_of_products_select_all_products_604.png

     
     # ./spec/support/request/authentication_helper.rb:14:in `login_as_admin_and_visit'
     # ./spec/system/admin/order_cycles/complex_editing_multiple_product_pages_spec.rb:21:in `block (3 levels) in <main>'
     # ./spec/system/support/cuprite_setup.rb:44:in `block (2 levels) in <main>'
     # -e:1:in `<main>'

Finished in 19.37 seconds (files took 1.13 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/system/admin/order_cycles/complex_editing_multiple_product_pages_spec.rb:37
```

**Comment by kirstenalarsen** [comment]
is this ready to move to test @RachL @filipefurtad0 ?

**Review by mkllnk — APPROVED** [review]

**Comment by filipefurtad0** [comment]
Hey @jibees , some conflicts on this branch :point_up_2:

**Comment by jibees** [comment]
> Hey @jibees , some conflicts on this branch 👇

Thanks for the ping ;) Should be ok now!

**Pull request merged**
